pkg/placement/controller.go

```diff
@@ -441,7 +442,46 @@ func (c *Controller) reconcile(ctx context.Context, key util.Key) error {
 	}
 
 	c.logger.Info("Delivering", "object", util.GenerateObjectInfoString(obj), "to clusters", clusters)
-	deliverObjectToManagedClusters(c.logger, c.ocmClient, obj, clusters, managedByPlacements, c.wdsName, singletonStatus)
+	for _, clName := range clusters {
+		// find which placement(s) select this managedCluster
+		placementNames := []string{}
+		for _, plName := range managedByPlacements {
+			plObj, err := c.getPlacementByName(plName)
+			if err != nil {
+				return err
+			}
+			pl, err := runtimeObjectToPlacement(plObj)
+			if err != nil {
+				return err
+			}
+			cl, err := ocm.GetClusterByName(c.ocmClient, clName)
+			if err != nil {
+				return err
+			}
+			// a placement selects a managedCluster iff the managedCluster is selected by every single selector
+			// i.e. selectors are ANDed
+			matches := true
+			for _, s := range pl.Spec.ClusterSelectors {
+				selector, err := metav1.LabelSelectorAsSelector(&s)
+				if err != nil {
+					return err
+				}
+				if !selector.Matches(labels.Set(cl.Labels)) {
+					matches = false
+					break
+				}
+			}
+			if matches {
+				placementNames = append(placementNames, plName)
+			}
+		}
+		manifest := ocm.WrapObject(obj)
+		util.SetManagedByPlacementLabels(manifest, c.wdsName, placementNames, singletonStatus)
+		err := reconcileManifest(c.ocmClient, manifest, clName)
```

Comment:
Do we need to do the above operations in case there were no matches at all and placementNames is empty ?

Reply:
Good question. My understanding is that we should always have a non-empty list because in order for each cluster to be included into the `clusters` list, there must be at least one placement(s) that select it.

Reply:
At this point there should be a non-empty list of clusters (that was checked before) and  a non empty list of managedByPlacements - since the latter is not checked it seems like a good idea to check

Reply:
Done in the latest push.